File: src/controller/ExpenseController.ts
import { NextApiRequest, NextApiResponse } from 'next';
import { ExpenseService } from '../service/ExpenseService';
import { CreateExpenseRequest, ExpenseQueryParams } from '../model/Expense';

export class ExpenseController {
  private expenseService: ExpenseService;

  constructor() {
    this.expenseService = new ExpenseService();
  }

  async createExpense(req: NextApiRequest, res: NextApiResponse): Promise<void> {
    try {
      const request: CreateExpenseRequest = req.body;
      const expense = await this.expenseService.createExpense(request);
      res.status(201).json(expense);
    } catch (error: any) {
      res.status(400).json({ error: error.message || 'Failed to create expense' });
    }
  }

  async getExpenses(req: NextApiRequest, res: NextApiResponse): Promise<void> {
    try {
      const queryParams: ExpenseQueryParams = {
        category: req.query.category as string,
        sort: req.query.sort as string
      };
      
      const expenses = await this.expenseService.getExpenses(queryParams);
      res.status(200).json(expenses);
    } catch (error: any) {
      res.status(500).json({ error: error.message || 'Failed to fetch expenses' });
    }
  }
}

File: backend/src/main/java/com/expensetracker/controller/ExpenseController.java
package com.expensetracker.controller;

import com.expensetracker.model.CreateExpenseRequest;
import com.expensetracker.model.Expense;
import com.expensetracker.service.ExpenseService;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/expenses")
@CrossOrigin(origins = "http://localhost:5173") // React dev server
public class ExpenseController {
    private final ExpenseService expenseService;

    public ExpenseController(ExpenseService expenseService) {
        this.expenseService = expenseService;
    }

    @PostMapping
    public ResponseEntity<?> createExpense(@Valid @RequestBody CreateExpenseRequest request) {
        try {
            Expense expense = expenseService.createExpense(request);
            return ResponseEntity.status(HttpStatus.CREATED).body(expense);
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                .body(new ErrorResponse(e.getMessage()));
        }
    }

    @GetMapping
    public ResponseEntity<List<Expense>> getExpenses(
            @RequestParam(required = false) String category,
            @RequestParam(required = false) String sort) {
        List<Expense> expenses = expenseService.getExpenses(category, sort);
        return ResponseEntity.ok(expenses);
    }

    // Error response class
    private static class ErrorResponse {
        private String error;

        public ErrorResponse(String error) {
            this.error = error;
        }

        public String getError() {
            return error;
        }

        public void setError(String error) {
            this.error = error;
        }
    }
}

